s3-streamer

```diff
@@ -0,0 +1,319 @@
+        with open(self.path(filename), 'wb+') as file:
+            file.write(data)
+
+    def delete(self, filenames):
+        for filename in filenames:
+            print(f'Delete {self.path(filename)}')
+            os.unlink(self.path(filename))
+
+
+class S3Destination(Destination):
+    def __init__(self, directory):
+        self.directory = directory.rstrip('/') + '/'
+
+    def url(self, filename):
+        return urllib.parse.urlparse(self.directory + filename)
+
+    def has(self, filename):
+        raise NotImplementedError('use Index')
+
+    def write(self, filename, data):
+        headers = {
+            'Content-Type': mimetypes.guess_type(filename)[0] or 'text/plain',
+            s3.ACL: s3.PUBLIC
+        }
+        with s3.urlopen(self.url(filename), data=data, method='PUT', headers=headers) as response:
+            print(response.status, filename)
+
+    def delete(self, filenames):
+        # to do: multi-object delete API
+        for filename in filenames:
+            with s3.urlopen(self.url(filename), method='DELETE') as response:
+                print(response)
+
+
+class Index(Destination):
+    def __init__(self, destination, filename='index.html'):
+        self.destination = destination
+        self.filename = filename
+        self.files = set()
+        self.dirty = True
+
+    def has(self, filename):
+        return filename in self.files
+
+    def write(self, filename, data):
+        self.destination.write(filename, data)
+        self.files.add(filename)
+        self.dirty = True
+
+    def delete(self, filenames):
+        self.destination.delete(self.destination, filenames)
+        self.files.difference_update(filenames)
+        self.dirty = True
+
+    def sync(self):
+        if self.dirty:
+            self.destination.write(self.filename, textwrap.dedent('''
+                <html>
+                  <body>
+                    <h1>Directory listing for /</h1>
+                    <hr>
+                    <ul>''' + ''.join(f'''
+                      <li><a href={f}>{f}</a></li> ''' for f in self.files) + '''
+                    </ul>
+                  </body>
+                </html>
+                ''').encode('utf-8'))
+            self.dirty = False
+
+
+class AttachmentsDirectory:
+    def __init__(self, destination, local_directory):
+        self.destination = destination
+        self.path = local_directory
+
+    def scan(self):
+        for entry in os.scandir(self.path):
+            if not self.destination.has(entry.name) and entry.is_file(follow_symlinks=False):
+                with open(entry.path, 'rb') as file:
+                    data = file.read()
+                self.destination.write(entry.name, data)
+
+
+class ChunkedUploader:
+    SIZE_LIMIT = 1000000  # 1MB
+    TIME_LIMIT = 10       # 10s
+
+    def __init__(self, index, filename, encoding=None):
+        self.input_decoder = codecs.getincrementaldecoder(encoding or locale.getpreferredencoding())()
```

Comment:
I'd rather not depend on the locale here. This has bitten us all too often in the past, with bots running under `C` locale and regurgitating UTF-8 input.

As I said, I'd rather drop that whole decode → encode dance completely and just handle byte streams, with breaking on the last `b\n` of a read block (as sending half a line isn't that useful anyway). This avoids potential locale crashes, hangs with internal buffers of the encoders, and saves CPU cycles on top of that. You seemed very eager to do the encode/decode, but let's at least do a fixed `utf-8` here.

Reply:
Let's chat about this.

My priorities here are:
 - it needs to be impossible to send a partial multi-byte character
 - i also want to never ever send non-utf8 data to the client.  that means we need to detect invalid sequences and transform them, anyway.
 - we should see all of the available output in the event of a hang.

The locale is going to be utf-8 in any reasonable universe.  I use the user's preferred locale out of a desire to be strictly correct.  I can turn that into an assert, if you like.